### PR TITLE
feat(provider/google): Add support for Shared VPC Networking (XPN).

### PR DESCRIPTION
**Pull request by duftler**

### CODE REVIEW

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/GCEUtil.groovy [hunk 1]
```diff
@@ -744,6 +736,21 @@ class GCEUtil {
...
+  public static String deriveProjectId(String fullUrl) {
+    if (!fullUrl) {
+      throw new IllegalArgumentException("Attempting to derive project id from empty resource url.")
+    }
+
+    List<String> urlParts = fullUrl.tokenize("/")
+    int indexOfProjectsToken = urlParts.indexOf("projects")
+
+    if (indexOfProjectsToken == -1) {
+      throw new IllegalArgumentException("Attempting to derive project id from resource url that does not contain 'projects'.")
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/GCEUtil.groovy]: maybe put the `fullUrl` in the exception for debugability.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy [hunk 2]
```diff
@@ -214,12 +214,14 @@ class CopyLastGoogleServerGroupAtomicOperation extends GoogleAtomicOperation<Dep
                                                      ancestorInstanceProperties.serviceAccounts)
 
       newDescription.network =
-        GCEUtil.getLocalName(description.network ?: ancestorInstanceProperties.networkInterfaces?.getAt(0)?.network)
+          description.network != null
+          ? description.network
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy]: :roll_eyes: :fishing_pole_and_fish:

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleNetworkCachingAgent.groovy [hunk 3]
```diff
@@ -45,9 +46,25 @@ class GoogleNetworkCachingAgent extends AbstractGoogleCachingAgent {
...
+    // TODO(duftler): Batch these 2 calls.
+    List<Network> networks =
+      timeExecute(compute.networks().list(project),
+                  "compute.networks.list", TAG_SCOPE, SCOPE_GLOBAL).items as List
+
+    if (xpnHostProject) {
+      List<Network> hostNetworks = timeExecute(compute.networks().list(xpnHostProject),
+                                               "compute.networks.list", TAG_SCOPE, SCOPE_GLOBAL).items as List
+
+      if (hostNetworks) {
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleNetworkCachingAgent.groovy]: `networks = (networks ?: []) + (hostNetworks ?: [])`

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleSecurityGroupCachingAgent.groovy [hunk 4]
```diff
@@ -196,9 +198,23 @@ class GoogleSecurityGroupCachingAgent extends AbstractGoogleCachingAgent impleme
...
-                         "compute.firewalls.list", TAG_SCOPE, SCOPE_GLOBAL
-      ).items as List<Firewall>
+      List<Firewall> firewalls = timeExecute(compute.firewalls().list(project),
+                                             "compute.firewalls.list", TAG_SCOPE, SCOPE_GLOBAL).items as List
+
+      if (xpnHostProject) {
+        List<Firewall> hostFirewalls = timeExecute(compute.firewalls().list(xpnHostProject),
+                                                   "compute.firewalls.list", TAG_SCOPE, SCOPE_GLOBAL).items as List
+
+        if (hostFirewalls) {
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleSecurityGroupCachingAgent.groovy]: Same groovification here as above

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleSubnetCachingAgent.groovy [hunk 5]
```diff
@@ -60,10 +61,26 @@ class GoogleSubnetCachingAgent extends AbstractGoogleCachingAgent {
...
+    List<Subnetwork> subnets = timeExecute(compute.subnetworks().list(project, region),
+                                           "compute.subnetworks.list",
+                                           TAG_SCOPE, SCOPE_REGIONAL, TAG_REGION, region).items as List
+
+    if (xpnHostProject) {
+      List<Subnetwork> hostSubnets = timeExecute(compute.subnetworks().list(xpnHostProject, region),
+                                                 "compute.subnetworks.list",
+                                                 TAG_SCOPE, SCOPE_REGIONAL, TAG_REGION, region).items as List
+
+      if (hostSubnets) {
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleSubnetCachingAgent.groovy]: And again

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/GCEUtil.groovy (hunk 1 above)
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/GCEUtil.groovy]: Done.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy (hunk 2 above)
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/CopyLastGoogleServerGroupAtomicOperation.groovy]: Still wrong...

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleNetworkCachingAgent.groovy (hunk 3 above)
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleNetworkCachingAgent.groovy]: Done.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleSecurityGroupCachingAgent.groovy (hunk 4 above)
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleSecurityGroupCachingAgent.groovy]: Done.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleSubnetCachingAgent.groovy (hunk 5 above)
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleSubnetCachingAgent.groovy]: Done.

**Comment by jawlitkp** [comment]
How to use or add shared VPC while installing spinnaker?

I am getting following error 

- Add the spinnaker-gce-account account
  Failure
Problems in default.provider.google:
! ERROR Network default not found via any configured google
  account.
- Failed to add account spinnaker-gce-account for provider
  google.

**Comment by duftler** [comment]
We discussed this offline and determined that there is no `default` network in their setup, and the bakery needs to be explicitly informed what network it should use (whether shared VPC or non-shared local network) or the validation step fails.

**Pull request merged**